### PR TITLE
Update audio and audioset_view to not rely on audio_view

### PR DESCRIPTION
**Pull request by stacimc**
⚠️ NOTE: The majority of the changes are to the audio sample data! Do not be alarmed by the line count.

## Description

The popularity refresh project aims to get rid of the audio and image matviews, and run data refreshes directly off the media tables. For audio, there are two problems:

* The `audioset_view`, which is also copied during a data refresh, is currently built from the `audio_view`.
* The API `audio` table has an `audio_set_foreign_identifier` column linking `audio` and `audio_set`. This column is not present on the catalog `audio` table, but rather a part of the `audio_set` json column. The column is only present in the `audio_view`, from which the ingestion server _currently_ copies directly into the API table.

So in order to run an audio data refresh from the audio table, and eventually get rid of the audio_view, we must:

* Add the `audio_set_foreign_identifier` column to the `audio` table. _This will allow the ingestion server to continue doing a simple bulk copy without having to transform the data, when we start doing the refresh from the audio table._
* Base the `audio_set` view off of `audio`, rather than `audio_view`. _This will allow us to drop the matview without losing the audioset information._
* Update ingestion to write to `audio_set_foreign_identifier`. _This ensures that newly ingested records populate audioset information correctly going forward._

This PR brings the _local_ environment up to speed. In production, we’ll need to do the following steps:

* In production DB, add the column (`ALTER TABLE audio ADD COLUMN audio_set_foreign_identifier character varying(1000)`)
* At this point, we can merge this PR. When audio provider DAGs run, they will correctly update new data.
* Backfill existing audio data. We can do this with SQL manually or use `batched_update` DAG:
        * `UPDATE audio SET audio_set_foreign_identifier = audio_set->>'foreign_identifier' WHERE audio_set IS NOT NULL;`

## Testing Instructions

We need to test both updates to the local testing environment (like my updates to the sample data csv), and the steps that will be run in production.

### Testing local updates

Start on `main` and run `just recreate` to ensure that you have clean sample data based on `main`.

Run `just catalog/pgcli` to open a pgcli session and verify the following:

```sql
> SELECT COUNT(*) FROM audio WHERE audio_set IS NOT NULL AND audio_set->>'foreign_identifier' IS NOT NULL;
+-------+
| count |
|-------|
| 673   |
+-------+
```

You can also inspect the first couple of `audio_set` dicts to manually cross-reference later:

```
> SELECT audio_set FROM audio WHERE audio_set IS NOT NULL LIMIT 10;
```

While you're all checked out on `main`, go ahead and also pull up an API response that includes audioset information (so you can compare to the results on this branch later). I looked at http://localhost:50280/v1/audio/?q=%27wish%20you%20were%20here%27.

Now checkout this branch and run `just recreate` to rebuild with the new schema and sample data. Note that under the covers this does a data refresh to populate the api with sample data, so we're also testing the data refresh in the process!

```sql
-- The correct number of rows have non-null audio_set_foreign_identifier
> SELECT COUNT(*) FROM audio WHERE audio_set_foreign_identifier IS NOT NULL;
+-------+
| count |
|-------|
| 673   |
+-------+

-- Verify the first 10 audio_sets look the same.
>  SELECT audio_set FROM audio WHERE audio_set IS NOT NULL LIMIT 10;
```

### Testing production

Checkout `main` again, run `just recreate` to ensure that you have clean sample data based on `main` once more.

Now checkout this branch again but do NOT run `just recreate`. Instead do the manual updates:

* `ALTER TABLE audio ADD COLUMN audio_set_foreign_identifier character varying(1000);`
* `UPDATE audio SET audio_set_foreign_identifier = audio_set->>'foreign_identifier' WHERE audio_set IS NOT NULL;`

And once again test that you have the correct number of rows with `audio_set_foreign_identifier` etc.

### Testing ingestion

Finally test that ingestion of audio providers works. Run a couple of audio DAGs and ensure they do not fail. I ran jamendo for ~2 minutes before manually marking the `pull_data` task as a success. Then:

```sql
-- Observe that we have new results that have the fid set
> SELECT COUNT(*) FROM audio WHERE audio_set_foreign_identifier IS NOT NULL;
+-------+
| count |
|-------|
| 1672  |
+-------+

-- The same number of results have non-null audio_set
> SELECT COUNT(*) FROM audio WHERE audio_set IS NOT NULL;
+-------+
| count |
|-------|
| 1672  |
+-------+
```

### Test data refresh

Run an audio data refresh and ensure it works as well.

### Test API response

For good measure, also double check that an actual API result matches what you got earlier on `main`. I looked at http://localhost:50280/v1/audio/?q=%27wish%20you%20were%20here%27 and noted its values in `audio_set_foreign_identifier` and `audioset`:

<details>
  <summary>It should look like this</summary>
  
  ```json
{
    "result_count": 1,
    "page_count": 1,
    "page_size": 20,
    "page": 1,
    "results": [
        {
            "id": "8624ba61-57f1-4f98-8a85-ece206c319cf",
            "title": "Wish You Were Here",
            "indexed_on": "2022-12-06T06:54:25Z",
            "foreign_landing_url": "https://www.jamendo.com/track/1214935",
            "url": "https://mp3d.jamendo.com/download/track/1214935/mp32",
            "creator": "The.madpix.project",
            "creator_url": "https://www.jamendo.com/artist/441585/the.madpix.project",
            "license": "by-nc-sa",
            "license_version": "3.0",
            "license_url": "https://creativecommons.org/licenses/by-nc-sa/3.0/",
            "provider": "jamendo",
            "source": "jamendo",
            "category": "music",
            "genres": [
                "dance",
                "electronic",
                "house"
            ],
            "filesize": 7139840,
            "filetype": "mp3",
            "tags": [
                {
                    "name": "vocal",
                    "accuracy": null
                },
                {
                    "name": "female",
                    "accuracy": null
                },
                {
                    "name": "speed_medium",
                    "accuracy": null
                },
                {
                    "name": "guitar",
                    "accuracy": null
                },
                {
                    "name": "strings",
                    "accuracy": null
                },
                {
                    "name": "energetic",
                    "accuracy": null
                },
                {
                    "name": "acoustic",
                    "accuracy": null
                },
                {
                    "name": "vocal",
                    "accuracy": null
                },
                {
                    "name": "voice",
                    "accuracy": null
                },
                {
                    "name": "funkyhouse",
                    "accuracy": null
                }
            ],
            "alt_files": null,
            "attribution": "\"Wish You Were Here\" by The.madpix.project is licensed under CC BY-NC-SA 3.0. To view a copy of this license, visit https://creativecommons.org/licenses/by-nc-sa/3.0/.",
            "fields_matched": [
                "title"
            ],
            "mature": false,
            "audio_set": {
                "title": "Wish You Were Here",
                "foreign_landing_url": "https://www.jamendo.com/album/145774/wish-you-were-here",
                "creator": "The.madpix.project",
                "creator_url": "https://www.jamendo.com/artist/441585/the.madpix.project",
                "url": "[https://usercontent.jamendo.com?type=album&id=145774&width=200](https://usercontent.jamendo.com/?type=album&id=145774&width=200)",
                "filesize": null,
                "filetype": null
            },
            "duration": 270000,
            "bit_rate": 128000,
            "sample_rate": 44100,
            "thumbnail": "http://localhost:50280/v1/audio/8624ba61-57f1-4f98-8a85-ece206c319cf/thumb/",
            "detail_url": "http://localhost:50280/v1/audio/8624ba61-57f1-4f98-8a85-ece206c319cf/",
            "related_url": "http://localhost:50280/v1/audio/8624ba61-57f1-4f98-8a85-ece206c319cf/related/",
            "waveform": "http://localhost:50280/v1/audio/8624ba61-57f1-4f98-8a85-ece206c319cf/waveform/",
            "unstable__sensitivity": []
        }
    ]
}
  ```
</details>

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by sarayourfriend** [comment]
@stacimc I will review this later today.

**Comment by sarayourfriend** [comment]
@stacimc I've gotten stuck running through the testing instructions. I haven't looked carefully at the code changes and am out of time for today, but I'll try and debug it further tomorrow. Feels like I'm missing a step somewhere or there might be a typo in that `UPDATE audio` query.

```
openledger> ALTER TABLE audio ADD COLUMN audio_set_foreign_identifier character varying(1000);
You're about to run a destructive command.
Do you want to proceed? (y/n): y
Your call!
ALTER TABLE
Time: 0.003s
openledger> UPDATE audio SET audio_set_foreign_identifier = audio_set->>’foreign_identifier’ WHERE audio_set IS NOT NULL;
You're about to run a destructive command.
Do you want to proceed? (y/n): 
You're about to run a destructive command.
Do you want to proceed? (y/n): y
Your call!
column "’foreign_identifier’" does not exist
LINE 1: ...io SET audio_set_foreign_identifier = audio_set->>’foreign_i...
                                                             ^
HINT:  Perhaps you meant to reference the column "audio.foreign_identifier".

Time: 0.010s
openledger>
```

**Comment by obulat** [comment]
> @stacimc I've gotten stuck running through the testing instructions in the "Test production" section. I haven't looked carefully at the code changes and am out of time for today, but I'll try and debug it further tomorrow. Feels like I'm missing a step somewhere or there might be a typo in that `UPDATE audio` query.
> 
> ```
> openledger> ALTER TABLE audio ADD COLUMN audio_set_foreign_identifier character varying(1000);
> You're about to run a destructive command.
> Do you want to proceed? (y/n): y
> Your call!
> ALTER TABLE
> Time: 0.003s
> openledger> UPDATE audio SET audio_set_foreign_identifier = audio_set->>’foreign_identifier’ WHERE audio_set IS NOT NULL;
> You're about to run a destructive command.
> Do you want to proceed? (y/n): 
> You're about to run a destructive command.
> Do you want to proceed? (y/n): y
> Your call!
> column "’foreign_identifier’" does not exist
> LINE 1: ...io SET audio_set_foreign_identifier = audio_set->>’foreign_i...
>                                                              ^
> HINT:  Perhaps you meant to reference the column "audio.foreign_identifier".
> 
> Time: 0.010s
> openledger>
> ```

The problem was caused by the single quotation marks mismatch: 

> The correct quotation mark for accessing keys in a JSON object in PostgreSQL is the single quote ('). You seem to be using a different kind of quotation marks (’), which is causing the error.

(reply from ChatGPT 😄  )

I'll edit the instructions.

**Review by obulat — CHANGES_REQUESTED** [review]
Thank you for very thorough testing instructions! I got stuck on the single quote issue, too, but after replacing it the local and production test steps went fine. Ingestion, however, did not write a `audio_set_foreign_identifier`, even though the `audio_set` is not `null`.

`select count(*) from audio where audio_set_foreign_identifier is not null`, I receive 0.

And this is what's saved in audio_set_foreign_identifier and audio_set:
```
>  select audio_set_foreign_identifier, audio_set from audio where audio_set is not null limit 1;

| <null> | {"title": "Ben... Mon cochon !!", "creator": "Rouler Pinder", "thumbnail": "https://usercontent.jamendo.com/?type=album&id=172&width=300", "creator_url": "https://www.jamendo.com/artist/155/rouler.pinder", "foreign_identifier": "172", "foreign_landing_url": "https://www.jamendo.com/album/172/ben-mon-cochon"} |
```

**Comment by stacimc** [comment]
Thanks @obulat -- I must have tested this alongside my other changes to remove the matview steps entirely, and missed this. The `audioset_view` was being recreated during the audio data refresh, because it used to be dropped as part of dropping the matview. Now that is no longer necessary, so I have removed that step.

Also apologies for the quote -- I drafted the PR instructions in Notes and copied them in, which resulted in the wrong character. Thanks for fixing Olga!

**Comment by obulat** [comment]
Thank you for the quick fix, @stacimc , I'll test the refresh steps tomorrow

**Comment by obulat** [comment]
Now the data refresh is stuck at `data_refresh.wait_for_create_and_populate_filtered_index`. Did I have to set something up to get it to work, @stacimc ?

On an unrelated note, I spent some time not understanding why I couldn't see the logs in Airflow. Turns out, I didn't update my env file after the airflow update. I wish there was a check that would say something like "Please update your env file with these variables" after any env.template changes.

**Comment by sarayourfriend** [comment]
Everything works for me aside from being able to run DAGs. I'm confident that this is something coming from my lack of understanding of the catalogue:

```
[2023-08-18, 07:50:56 UTC] {taskinstance.py:1431} INFO - TSV file /var/workflow_output/jamendo_audio_v001_20230818075052.tsv does not exist.
[2023-08-18, 07:50:56 UTC] {taskinstance.py:1345} INFO - Marking task as SKIPPED. dag_id=jamendo_workflow, task_id=ingest_data.load_audio_data.copy_to_s3, execution_date=20230818T075051, start_date=20230818T075055, end_date=20230818T075056
```

These are logs from the copy_to_s3 task in the Jamendo workflow. At a glance I haven't found any docs that describe how to work around this :confused: Is there some bootstraping I need to do aside from `just init`?

**Comment by stacimc** [comment]
@obulat There should be no special set-up to run the filtered index creation. Did you remember to turn on the `create_audio_filtered_index` DAG (asking because I forget all the time 😄)? I tested again and it works for me.

@sarayourfriend  I'm not sure what's going on there 😮 It's possible your issue might be related to the fact that the Airflow 2.6.3 upgrade got pushed to `main`. I've rebased this branch so both branches are on the same Airflow version now. As Olga mentioned, for the Airflow upgrade you'll need to make an update to your `catalog/.env` file (relevant [commit](https://github.com/WordPress/openverse/pull/2748/files#diff-fc01a83b2fe674b9efd5392da75a6941f27e8e3acfb56aefbe1aa4e767751784R49), tldr you need to urlencode the AWS connection var, so `AIRFLOW_CONN_AWS_DEFAULT=aws://test_key:test_secret@?region_name=us-east-1&endpoint_url=http%3A%2F%2Fs3%3A5000`). After that I would try a full `just recreate` on this branch.

**Comment by obulat** [comment]
Of course, I forgot to turn on the filtered dag :😀 I'll try again later, I'm almost there!

**Comment by sarayourfriend** [comment]
Hmm, I was running off of a `main` that didn't have the updated Airflow yet, so I don't think it was the version. I've tried it again from scratch, checking out `main`, `just recreate`, then switching to the branch, running the SQL commands, then trying to run Jamendo and nothing happens :thinking: The pull data step passes immediately with zero records to load.

**Review by obulat — APPROVED** [review]
I managed to run all of the steps from the testing instructions.

There was a problem with the `audioset_view` at `create_materialized_popularity_view` step:
*DAG*: `audio_data_refresh` *Task*: `recreate_matview.create_materialized_popularity_view` *Logical Date*: 2023-08-21T06:31:48Z *Log*: http://localhost:8080/log?execution_date=2023-08-21T06%3A31%3A48.209091%2B00%3A00&task_id=recreate_matview.create_materialized_popularity_view&dag_id=audio_data_refresh&map_index=-1 *Exception*: relation "audioset_view" already exists *Exception Type*: `psycopg2.errors.DuplicateTable`

When I manually deleted `audioset_view` and restarted the DAG, it all worked. I guess it's a problem with my local setup, though.

I had to update the ES prod connection ID and `ES_INDEX_READINESS_RECORD_COUNT` (lower to 5000) in the `env` file to make it all work.

**Review by krysal — COMMENTED** [review]
Generally, everything looks good to me in the files changed. thanks for the detailed explanation! I left a pair of questions for clarification.

File: catalog/dags/common/storage/tsv_columns.py [hunk 1]
```diff
@@ -39,6 +39,7 @@
             col.AUDIO_SET,
             col.SET_POSITION,
             col.ALT_FILES,
+            col.AUDIO_SET_FOREIGN_IDENTIFIER,
```
[the remark this answers is not in the document]
Comment (obulat) [reply catalog/dags/common/storage/tsv_columns.py:42]: Right, @krysal! We should increment the columns version to make sure that we can still ingest old TSVs, @stacimc.

[the remark this answers is not in the document]
Comment (stacimc) [reply catalog/dags/common/storage/tsv_columns.py:71]: *Great* catch -- incremented the version!

File: catalog/dags/common/popularity/sql.py [hunk 2]
```diff
@@ -435,8 +432,6 @@ def create_media_view(
     )
     postgres.run(create_view_query)
     postgres.run(add_idx_query)
-    if media_type == AUDIO:
-        postgres.run(create_audioset_view_query())
```
[the remark this answers is not in the document]
Comment (stacimc) [reply catalog/dags/common/popularity/sql.py:439]: I thought we might need it for the full sql recreate DAGs, but confirmed those will be going away now -- so, not anymore! Removed :)

**Comment by stacimc** [comment]
> There was a problem with the audioset_view at create_materialized_popularity_view step:

@obulat, are you sure you have the latest changes from this PR pulled? That error should not be possible, because that task no longer attempts to create the audioset_view (relevant [change](https://github.com/WordPress/openverse/pull/2803/files#diff-ac660c0e19ecbde9ea83756b012868c62e9e4478001ec2df56d075df779d8c7fL439)).

**Comment by obulat** [comment]
> > There was a problem with the audioset_view at create_materialized_popularity_view step:
> 
> @obulat, are you sure you have the latest changes from this PR pulled? That error should not be possible, because that task no longer attempts to create the audioset_view (relevant [change](https://github.com/WordPress/openverse/pull/2803/files#diff-ac660c0e19ecbde9ea83756b012868c62e9e4478001ec2df56d075df779d8c7fL439)).

I probably didn't have the latest version. I'll check out the latest version tomorrow and let you know how it goes.

**Comment by stacimc** [comment]
> trying to run Jamendo and nothing happens 🤔 The pull data step passes immediately with zero records to load.

@sarayourfriend can you verify that you have the API key for Jamendo set in your `catalog/.env`? That's the only other thing I can think of -- I get results from Jamendo very quickly. If that's not it and you have some time to test, I would be curious if you get results from Jamendo when pulling on `main`? Unfortunately Jamendo is definitely our easiest/quickest testing option for audio.

**Comment by obulat** [comment]
I re-tried running the audio refresh - and everything worked well, @stacimc :)

**Review by sarayourfriend — APPROVED** [review]
Finally was able to work through all the testing (I needed my own personal Jamendo key, the one in their docs for read-only testing _does not work_ for our use-case apparently!) and it all works perfectly :tada:

**Pull request merged**